### PR TITLE
[core] refactor disconnect message processing and enrich WorkExitType

### PR DESCRIPTION
**Pull request by oliverhu**
## Why are these changes needed?
We have two client disconnect messages currently and the "intentional" keyword is confusing. The PR adds explicit exit type to distinguish different worker exit type thus we could handle client disconnect with more clarity. 

## Related issue number
https://github.com/ray-project/ray/pull/13321 would depend on this.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/protobuf/common.proto [hunk 1]
```diff
@@ -442,3 +442,15 @@ message MetricPoint {
...
+enum WorkerExitType {
+  // Worker exit due to system level failures (i.e. worker crash).
+  SYSTEM_ERROR_EXIT = 0;
+  // Intended, initiated worker exit via raylet API.
+  INTENDED_EXIT = 1;
+  // Worker exit due to resource bundle release.
+  UNUSED_RESOURCE_RELEASED = 2;
+  // Worker exit due to placement group removal.
+  PLACEMENT_GROUP_REMOVED = 3;
+}
```
Comment (ffbin) [inline src/ray/protobuf/common.proto:456]: Add a new line.

File: src/ray/raylet/node_manager.h [hunk 2]
```diff
@@ -741,6 +743,9 @@ class NodeManager : public rpc::NodeManagerServiceHandler,
   void OnObjectMissing(const ObjectID &object_id,
                        const std::vector<TaskID> &waiting_task_ids) override;
 
+  void DisconnectClient(
```
Comment (ffbin) [inline src/ray/raylet/node_manager.h:746]: How about add function description?

File: src/ray/raylet/format/node_manager.fbs [hunk 3]
```diff
@@ -38,9 +38,6 @@ enum MessageType:int {
   // Notify the raylet that this client is disconnecting unexpectedly.
```
Comment (ffbin) [inline src/ray/raylet/format/node_manager.fbs:38]: Does the description need to be modified?

**Review by ffbin — APPROVED** [review]
LGTM, just some small comments.

**Review by rkooo567 — APPROVED** [review]

**Pull request merged**